// @atom-ide-community/nuclide-commons-ui/BlockDecoration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactDom = _interopRequireDefault(require("react-dom"));

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class BlockDecoration extends React.Component {
  constructor(props) {
    super(props);
    this._marker = void 0;
    this._item = void 0;
    this._item = document.createElement('div');
  }

  componentDidMount() {
    this._createMarker();
  }

  componentWillUnmount() {
    this._marker.destroy();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.editor !== this.props.editor) {
      this._createMarker();
    } else if (!prevProps.range.isEqual(this.props.range)) {
      this._marker.setBufferRange(this.props.range);
    }
  }

  _createMarker() {
    if (this._marker) {
      this._marker.destroy();
    }

    const {
      range,
      editor
    } = this.props;
    this._marker = editor.markBufferRange(range);
    editor.decorateMarker(this._marker, {
      type: 'block',
      position: 'after',
      item: this._item
    });
  }

  render() {
    return /*#__PURE__*/_reactDom.default.createPortal(this.props.children, this._item);
  }

}

exports.default = BlockDecoration;
