// @atom-ide-community/nuclide-commons-ui/Button.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Button = require("./Button");

var _ButtonGroup = require("./ButtonGroup");

var _ButtonToolbar = require("./ButtonToolbar");

var _Block = require("./Block");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const ButtonSizeExample = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Button.Button, {
  tooltip: {
    title: 'An Extra Small Tooltip'
  },
  className: "inline-block",
  size: "EXTRA_SMALL"
}, "extra_small"), /*#__PURE__*/React.createElement(_Button.Button, {
  tooltip: {
    title: 'A Small Tooltip'
  },
  className: "inline-block",
  size: "SMALL"
}, "small"), /*#__PURE__*/React.createElement(_Button.Button, {
  tooltip: {
    title: 'A Regular Sized Tooltip'
  },
  className: "inline-block"
}, "regular"), /*#__PURE__*/React.createElement(_Button.Button, {
  tooltip: {
    title: 'A Large Tooltip'
  },
  className: "inline-block",
  size: "LARGE"
}, "large"));

const ButtonDisabledExample = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Button.Button, {
  className: "inline-block"
}, "enabled"), /*#__PURE__*/React.createElement(_Button.Button, {
  className: "inline-block",
  disabled: true
}, "disabled"));

const ButtonColorExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, null, /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "PRIMARY"
}, "primary"), /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "INFO"
}, "info"), /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "SUCCESS"
}, "success"), /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "WARNING"
}, "warning"), /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "ERROR"
}, "error"))), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement("p", null, "selected:"), /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, null, /*#__PURE__*/React.createElement(_Button.Button, {
  selected: true,
  buttonType: "PRIMARY"
}, "primary"), /*#__PURE__*/React.createElement(_Button.Button, {
  selected: true,
  buttonType: "INFO"
}, "info"), /*#__PURE__*/React.createElement(_Button.Button, {
  selected: true,
  buttonType: "SUCCESS"
}, "success"), /*#__PURE__*/React.createElement(_Button.Button, {
  selected: true,
  buttonType: "WARNING"
}, "warning"), /*#__PURE__*/React.createElement(_Button.Button, {
  selected: true,
  buttonType: "ERROR"
}, "error"))));

const ButtonIconExample = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, null, /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "gear"
}), /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "cloud-download"
}), /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "code"
}), /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "check"
}), /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "device-mobile"
}), /*#__PURE__*/React.createElement(_Button.Button, {
  icon: "alert"
})));

const ButtonGroupExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, {
  size: "EXTRA_SMALL"
}, /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "SUCCESS"
}, "extra small"), /*#__PURE__*/React.createElement(_Button.Button, null, "button"), /*#__PURE__*/React.createElement(_Button.Button, null, "group"))), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, {
  size: "SMALL"
}, /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "SUCCESS"
}, "small"), /*#__PURE__*/React.createElement(_Button.Button, null, "button"), /*#__PURE__*/React.createElement(_Button.Button, null, "group"))), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, null, /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "SUCCESS"
}, "regular"), /*#__PURE__*/React.createElement(_Button.Button, null, "button"), /*#__PURE__*/React.createElement(_Button.Button, null, "group"))), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, {
  size: "LARGE"
}, /*#__PURE__*/React.createElement(_Button.Button, {
  buttonType: "SUCCESS"
}, "large"), /*#__PURE__*/React.createElement(_Button.Button, null, "button"), /*#__PURE__*/React.createElement(_Button.Button, null, "group"))));

const ButtonToolbarExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ButtonToolbar.ButtonToolbar, null, /*#__PURE__*/React.createElement(_ButtonGroup.ButtonGroup, null, /*#__PURE__*/React.createElement(_Button.Button, null, "ButtonGroup"), /*#__PURE__*/React.createElement(_Button.Button, null, "in a"), /*#__PURE__*/React.createElement(_Button.Button, null, "toolbar")), /*#__PURE__*/React.createElement(_Button.Button, null, "single buttons"), /*#__PURE__*/React.createElement(_Button.Button, null, "in toolbar"))));

const ButtonExamples = {
  sectionName: 'Buttons',
  description: 'For clicking things.',
  examples: [{
    title: 'Button sizes',
    component: ButtonSizeExample
  }, {
    title: 'Disabled/enabled',
    component: ButtonDisabledExample
  }, {
    title: 'Button colors',
    component: ButtonColorExample
  }, {
    title: 'Buttons with icons',
    component: ButtonIconExample
  }, {
    title: 'Button Group',
    component: ButtonGroupExample
  }, {
    title: 'Button Toolbar',
    component: ButtonToolbarExample
  }]
};
exports.ButtonExamples = ButtonExamples;
