// @atom-ide-community/nuclide-commons-ui/Checkbox.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Checkbox = require("./Checkbox");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const NOOP = () => {};

const CheckboxExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Checkbox.Checkbox, {
  checked: false,
  onClick: NOOP,
  onChange: NOOP,
  label: "A Checkbox."
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Checkbox.Checkbox, {
  onClick: NOOP,
  onChange: NOOP,
  checked: true,
  label: "A checked Checkbox."
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Checkbox.Checkbox, {
  onClick: NOOP,
  onChange: NOOP,
  disabled: true,
  checked: false,
  label: "A disabled Checkbox."
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Checkbox.Checkbox, {
  onClick: NOOP,
  onChange: NOOP,
  checked: true,
  disabled: true,
  label: "A disabled, checked Checkbox."
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Checkbox.Checkbox, {
  onClick: NOOP,
  onChange: NOOP,
  indeterminate: true,
  checked: false,
  label: "An indeterminate Checkbox."
})));

const CheckboxExamples = {
  sectionName: 'Checkbox',
  description: '',
  examples: [{
    title: '',
    component: CheckboxExample
  }]
};
exports.CheckboxExamples = CheckboxExamples;
