// @atom-ide-community/nuclide-commons-ui/DateSelector.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateSelectorExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _AtomInput = require("./AtomInput");

var _Block = require("./Block");

var _DateSelector = require("./DateSelector");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const ONE_WEEK = 1000 * 60 * 60 * 24 * 7;

class BasicExample extends React.Component {
  constructor(props) {
    super(props);

    this._onChangeDate = date => {
      this.setState({
        date
      });
    };

    this.state = {
      date: new Date()
    };
  }

  render() {
    return /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_DateSelector.DateSelector, {
      start: this.state.date,
      earliestDate: new Date(Date.now() - ONE_WEEK),
      latestDate: new Date(Date.now() + ONE_WEEK),
      onDatesChange: this._onChangeDate
    }));
  }

}

class RangeExample extends React.Component {
  constructor(props) {
    super(props);

    this._onChangeDate = (start, end) => {
      this.setState({
        start,
        end
      });
    };

    this.state = {
      start: null,
      end: null
    };
  }

  render() {
    return /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_DateSelector.DateSelector, {
      start: this.state.start,
      end: this.state.end,
      latestDate: new Date(Date.now() + ONE_WEEK),
      onDatesChange: this._onChangeDate,
      selectRange: true
    }));
  }

}

class OpenOnFocusExample extends React.Component {
  constructor(props) {
    super(props);

    this._onChangeDate = date => {
      this.setState({
        date
      });
    };

    this.state = {
      date: new Date()
    };
  }

  render() {
    return /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '27em'
      }
    }, /*#__PURE__*/React.createElement(_DateSelector.DateSelector, {
      start: this.state.date,
      onDatesChange: this._onChangeDate,
      showOnlyOnFocus: true
    }, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
      value: this.state.date ? this.state.date.toDateString() : '',
      placeholderText: "selected date"
    }))));
  }

}

const DateSelectorExamples = {
  sectionName: 'DateSelector',
  description: 'Renders a date selector.',
  examples: [{
    title: 'Select single date example',
    component: BasicExample
  }, {
    title: 'Select range of dates example',
    component: RangeExample
  }, {
    title: 'Open calendar on focus example',
    component: OpenOnFocusExample
  }]
};
exports.DateSelectorExamples = DateSelectorExamples;
