// @atom-ide-community/nuclide-commons-ui/DraggableFile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dragEventCameFromDraggableFile = dragEventCameFromDraggableFile;
exports.default = void 0;

var _UniversalDisposable = _interopRequireDefault(require("@atom-ide-community/nuclide-commons/UniversalDisposable"));

var React = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _rxjsCompatUmdMin = require("rxjs-compat/bundles/rxjs-compat.umd.min.js");

var _analytics = require("@atom-ide-community/nuclide-commons/analytics");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const MAGIC_DATA_TRANSFER_KEY = 'nuclide-draggable-file';

class DraggableFile extends React.Component {
  constructor(...args) {
    super(...args);
    this._disposables = void 0;

    this._onDragStart = e => {
      const {
        dataTransfer
      } = e;

      if (dataTransfer != null) {
        dataTransfer.effectAllowed = 'move';
        dataTransfer.setData('initialPath', this.props.uri); // For security reasons, dragEnter events can't see the values of
        // `dataTransfer`, just the keys. So, we use a unique key to enable other
        // components to infer if a drag event contains a draggable file.

        dataTransfer.setData(MAGIC_DATA_TRANSFER_KEY, ''); // Allow draggable files to be dragged into the tab bar.

        dataTransfer.setData('text/plain', this.props.uri);
        dataTransfer.setData('atom-event', 'true');
        dataTransfer.setData('allow-all-locations', 'true');
        (0, _analytics.track)('draggable-file:drag-start', {
          source: this.props.trackingSource,
          uri: this.props.uri
        });
      }
    };
  }

  componentDidMount() {
    const el = _reactDom.default.findDOMNode(this);

    this._disposables = new _UniversalDisposable.default( // Because this element can be inside of an Atom panel (which adds its own drag and drop
    // handlers) we need to sidestep React's event delegation.
    _rxjsCompatUmdMin.Observable.fromEvent(el, 'dragstart').subscribe(this._onDragStart));
  }

  componentWillUnmount() {
    this._disposables.dispose();
  }

  render() {
    const {
      uri,
      trackingSource,
      draggable = true,
      ...restProps
    } = this.props; // https://discuss.atom.io/t/drag-drop/21262/14

    const tabIndex = -1;
    return /*#__PURE__*/React.createElement("div", _extends({
      draggable: draggable,
      tabIndex: tabIndex
    }, restProps));
  }

}

exports.default = DraggableFile;

function dragEventCameFromDraggableFile(event) {
  const {
    dataTransfer
  } = event;

  if (dataTransfer == null) {
    return false;
  }

  return dataTransfer.types.includes(MAGIC_DATA_TRANSFER_KEY);
}
