// @atom-ide-community/nuclide-commons-ui/FilterReminder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class FilterReminder extends React.Component {
  constructor(...args) {
    super(...args);

    this.handleClick = e => {
      e.preventDefault();
      this.props.onReset();
    };
  }

  render() {
    var _this$props$noun, _this$props$nounPlura;

    const {
      filteredRecordCount
    } = this.props;

    if (filteredRecordCount === 0) {
      return null;
    }

    const noun = (_this$props$noun = this.props.noun) !== null && _this$props$noun !== void 0 ? _this$props$noun : 'item';
    const nounPlural = (_this$props$nounPlura = this.props.nounPlural) !== null && _this$props$nounPlura !== void 0 ? _this$props$nounPlura : `${noun}s`;
    return /*#__PURE__*/React.createElement("div", {
      className: "nuclide-filter-reminder"
    }, /*#__PURE__*/React.createElement("div", {
      className: "nuclide-filter-reminder-message"
    }, /*#__PURE__*/React.createElement("pre", null, filteredRecordCount, ' ', filteredRecordCount === 1 ? `${noun} is` : `${nounPlural} are`, ' ', "hidden by filters.")), /*#__PURE__*/React.createElement("a", {
      href: "#",
      onClick: this.handleClick
    }, /*#__PURE__*/React.createElement("pre", null, "Show all ", nounPlural)));
  }

}

exports.default = FilterReminder;
