// @atom-ide-community/nuclide-commons-ui/FullWidthProgressBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var _classnames = _interopRequireDefault(require("classnames"));

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class FullWidthProgressBar extends React.Component {
  render() {
    const className = (0, _classnames.default)('nuclide-ui-full-width-progress-bar');
    return /*#__PURE__*/React.createElement("div", {
      className: className,
      hidden: this.props.visible === false
    }, this._renderBar());
  }

  _isIndeterminate() {
    return this.props.progress == null;
  }

  _renderBar() {
    if (this._isIndeterminate()) {
      return /*#__PURE__*/React.createElement("div", {
        className: "indeterminate"
      });
    }

    (0, _assert.default)(this.props.progress != null);
    return /*#__PURE__*/React.createElement(Bar, {
      progress: this.props.progress
    });
  }

}

exports.default = FullWidthProgressBar;

class Bar extends React.Component {
  render() {
    const pct = Math.max(0, Math.min(100, this.props.progress * 100));
    return /*#__PURE__*/React.createElement("div", {
      className: "nuclide-ui-full-width-progress-bar-bar",
      style: {
        width: `${pct}%`
      }
    });
  }

}
