// @atom-ide-community/nuclide-commons-ui/FullWidthProgressBar.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FullWidthProgressBarExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _FullWidthProgressBar = _interopRequireDefault(require("./FullWidthProgressBar"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const Wrapper = ({
  children
}) => /*#__PURE__*/React.createElement("div", {
  style: {
    position: 'relative',
    paddingBottom: 5
  }
}, children);

const FullWidthProgressBarExample = () => /*#__PURE__*/React.createElement("div", null, "0%:", /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(Wrapper, null, /*#__PURE__*/React.createElement(_FullWidthProgressBar.default, {
  progress: 0,
  visible: true
}))), "50%:", /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(Wrapper, null, /*#__PURE__*/React.createElement(_FullWidthProgressBar.default, {
  progress: 0.5,
  visible: true
}))), "100%:", /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(Wrapper, null, /*#__PURE__*/React.createElement(_FullWidthProgressBar.default, {
  progress: 1,
  visible: true
}))), "Indeterminate (progress=null):", /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(Wrapper, null, /*#__PURE__*/React.createElement(_FullWidthProgressBar.default, {
  progress: null,
  visible: true
}))));

const FullWidthProgressBarExamples = {
  sectionName: 'FullWidthProgressBar',
  description: 'A subtle progress indicator that stretches across an entire pane or panel, indicating general progress.',
  examples: [{
    title: 'FullWidthProgressBar',
    component: FullWidthProgressBarExample
  }]
};
exports.FullWidthProgressBarExamples = FullWidthProgressBarExamples;
