// @atom-ide-community/nuclide-commons-ui/Highlight.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HighlightExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Highlight = require("./Highlight");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const HighlightExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Highlight.Highlight, null, "Default")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Highlight.Highlight, {
  color: _Highlight.HighlightColors.info
}, "Info")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Highlight.Highlight, {
  color: _Highlight.HighlightColors.success
}, "Success")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Highlight.Highlight, {
  color: _Highlight.HighlightColors.warning
}, "Warning")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Highlight.Highlight, {
  color: _Highlight.HighlightColors.error
}, "Error")));

const HighlightExamples = {
  sectionName: 'Highlight',
  description: 'Highlights are useful for calling out inline content, such as tags.',
  examples: [{
    title: 'Highlights',
    component: HighlightExample
  }]
};
exports.HighlightExamples = HighlightExamples;
