// @atom-ide-community/nuclide-commons-ui/Link.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _event = require("@atom-ide-community/nuclide-commons/event");

var _UniversalDisposable = _interopRequireDefault(require("@atom-ide-community/nuclide-commons/UniversalDisposable"));

var React = _interopRequireWildcard(require("react"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * A more accessible abstraction over `<a>`s in Atom, where the default behavior
 * prevents pressing Enter/Return from opening a link. In Atom, the `native-key-bindings`
 * on `<a>`s addresses this, but prevents tabbing to/from these elements.
 *
 * Instead, listen for 'atom:core-confirm' (which by default occurs on Enter/Return)
 * and dispatch a simulated click MouseEvent when these happen. This is not unlike
 * typical browser behavior when Enter/Return is pressed on links, triggering their
 * 'click' event with a simulated MouseEvent.
 */
class Link extends React.Component {
  constructor(...args) {
    super(...args);
    this._disposable = void 0;
    this._node = void 0;

    this._updateNode = node => {
      this._node = node;
    };
  }

  componentDidMount() {
    this._disposable = new _UniversalDisposable.default((0, _event.observableFromSubscribeFunction)(cb => atom.commands.add((0, _nullthrows.default)(this._node), 'core:confirm', cb)).subscribe(() => {
      if (this._node == null) {
        return;
      }

      this._node.dispatchEvent(new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
        view: window
      }));
    }));
  }

  componentWillUnmount() {
    if (this._disposable != null) {
      this._disposable.dispose();
    }
  }

  render() {
    return /*#__PURE__*/React.createElement("a", _extends({}, this.props, {
      ref: this._updateNode
    }));
  }

}

exports.default = Link;
