// @atom-ide-community/nuclide-commons-ui/Message.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Message = exports.MessageTypes = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const MessageTypes = Object.freeze({
  default: 'default',
  info: 'info',
  success: 'success',
  warning: 'warning',
  error: 'error'
});
exports.MessageTypes = MessageTypes;
const MessageTypeClassNames = Object.freeze({
  default: 'nuclide-ui-message-default',
  error: 'nuclide-ui-message-error',
  info: 'nuclide-ui-message-info',
  success: 'nuclide-ui-message-success',
  warning: 'nuclide-ui-message-warning'
});

const Message = props => {
  const {
    className,
    children,
    type
  } = props;
  const resolvedType = type == null ? MessageTypes.default : type;
  const newClassName = (0, _classnames.default)(className, 'nuclide-ui-message', MessageTypeClassNames[resolvedType]);
  return /*#__PURE__*/React.createElement("div", {
    className: newClassName
  }, children);
};

exports.Message = Message;
