// @atom-ide-community/nuclide-commons-ui/Message.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Message = require("./Message");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const MessageExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Message.Message, null, /*#__PURE__*/React.createElement("h2", null, "Message"), "Hello, I'm a simple message.")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Message.Message, {
  type: _Message.MessageTypes.info
}, "Hello I'm an ", /*#__PURE__*/React.createElement("strong", null, "info"), " message.")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Message.Message, {
  type: _Message.MessageTypes.success
}, "Hello I'm a ", /*#__PURE__*/React.createElement("strong", null, "success"), " message.")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Message.Message, {
  type: _Message.MessageTypes.warning
}, "Hello I'm a ", /*#__PURE__*/React.createElement("strong", null, "warning"), " message.")), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Message.Message, {
  type: _Message.MessageTypes.error
}, "Hello I'm an ", /*#__PURE__*/React.createElement("strong", null, "error"), " message.")));

const MessageExamples = {
  sectionName: 'Messages',
  description: 'Message boxes are used to surface issues, such as warnings, inline within Nuclide.',
  examples: [{
    title: 'Basic Messages',
    component: MessageExample
  }]
};
exports.MessageExamples = MessageExamples;
