// @atom-ide-community/nuclide-commons-ui/PanelComponentScroller.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PanelComponentScroller = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class PanelComponentScroller extends React.Component {
  render() {
    const style = this.props.overflowX == null ? null : {
      overflowX: this.props.overflowX
    };
    const className = (0, _classnames.default)(this.props.className, 'nuclide-ui-panel-component-scroller', 'nuclide-scrollbar-style-fix', {
      'nuclide-ui-panel-component-scroller--column': this.props.flexDirection === 'column'
    });
    return (
      /*#__PURE__*/
      // $FlowFixMe(>=0.53.0) Flow suppress
      React.createElement("div", {
        className: className,
        style: style,
        onScroll: this.props.onScroll,
        onFocus: this.props.onFocus
      }, this.props.children)
    );
  }

}

exports.PanelComponentScroller = PanelComponentScroller;
