// @atom-ide-community/nuclide-commons-ui/ProgressIndicators.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressIndicatorExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _ProgressBar = require("./ProgressBar");

var _LoadingSpinner = require("./LoadingSpinner");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const ProgressBarExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ProgressBar.ProgressBar, null)), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ProgressBar.ProgressBar, {
  max: 100,
  value: 0
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ProgressBar.ProgressBar, {
  max: 100,
  value: 50
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_ProgressBar.ProgressBar, {
  max: 100,
  value: 100
})));

const LoadingSpinnerExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_LoadingSpinner.LoadingSpinner, {
  size: "EXTRA_SMALL"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_LoadingSpinner.LoadingSpinner, {
  size: "SMALL"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_LoadingSpinner.LoadingSpinner, {
  size: "MEDIUM"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_LoadingSpinner.LoadingSpinner, {
  size: "LARGE"
})));

const ProgressIndicatorExamples = {
  sectionName: 'Progress Indicators',
  description: 'Show that work is being performed. Consider using one of these for any work > 1s.',
  examples: [{
    title: 'ProgressBar',
    component: ProgressBarExample
  }, {
    title: 'LoadingSpinner',
    component: LoadingSpinnerExample
  }]
};
exports.ProgressIndicatorExamples = ProgressIndicatorExamples;
