// @atom-ide-community/nuclide-commons-ui/PulseButton.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PulseButtonExample = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _PulseButtonWithTooltip = _interopRequireDefault(require("./PulseButtonWithTooltip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class Example extends React.Component {
  render() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement("div", {
      style: {
        height: 100,
        width: '100%',
        display: 'flex'
      }
    }, /*#__PURE__*/React.createElement(_PulseButtonWithTooltip.default, {
      ariaLabel: "New feature!",
      wrapperStyle: {
        margin: 'auto'
      },
      tooltipText: "Look I'm a tooltip!"
    }))));
  }

}

const PulseButtonExample = {
  sectionName: 'PulseButton',
  description: 'A glowing button that often triggers a dismissable tooltip',
  examples: [{
    title: 'PulseButton',
    component: Example
  }]
};
exports.PulseButtonExample = PulseButtonExample;
