// @atom-ide-community/nuclide-commons-ui/RadioGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
// Globally unique ID used as the "name" attribute to group radio inputs.
let uid = 0;

/**
 * A managed radio group component. Accepts arbitrary React elements as labels.
 */
class RadioGroup extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      uid: uid++
    };
  }

  render() {
    const {
      className,
      onSelectedChange,
      optionLabels,
      selectedIndex
    } = this.props;
    const checkboxes = optionLabels.map((labelContent, i) => {
      const id = 'nuclide-radiogroup-' + uid + '-' + i;
      return /*#__PURE__*/React.createElement("div", {
        key: i,
        className: "nuclide-ui-radiogroup-div"
      }, /*#__PURE__*/React.createElement("input", {
        className: "input-radio",
        type: "radio",
        checked: i === selectedIndex,
        name: 'radiogroup-' + this.state.uid,
        id: id,
        onChange: () => {
          onSelectedChange(i);
        }
      }), /*#__PURE__*/React.createElement("label", {
        className: "input-label nuclide-ui-radiogroup-label",
        htmlFor: id
      }, labelContent));
    });
    return /*#__PURE__*/React.createElement("div", {
      className: className
    }, checkboxes);
  }

}

exports.default = RadioGroup;
RadioGroup.defaultProps = {
  optionLabels: [],
  onSelectedChange: selectedIndex => {},
  selectedIndex: 0
};
