// @atom-ide-community/nuclide-commons-ui/ReactMountRootElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var React = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _getDisplayName = _interopRequireDefault(require("@atom-ide-community/nuclide-commons/getDisplayName"));

var _trackReactProfilerRender = _interopRequireDefault(require("@atom-ide-community/nuclide-commons/trackReactProfilerRender"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

/* eslint-env browser */
// $FlowFixMe Profiler is neither stable nor typed
const Profiler = React.unstable_Profiler;
/**
 * A custom HTMLElement we render React elements into.
 */

class ReactMountRootElement extends HTMLElement {
  constructor(...args) {
    super(...args);
    this._reactElement = void 0;
    this._profileName = void 0;
  }

  setReactElement(reactElement, profileName) {
    this._reactElement = reactElement;
    this._profileName = profileName;
  }

  connectedCallback() {
    var _this$_profileName;

    if (this._reactElement == null) {
      return;
    }

    _reactDom.default.render( /*#__PURE__*/React.createElement(Profiler, {
      id: `RootElement(${(_this$_profileName = this._profileName) !== null && _this$_profileName !== void 0 ? _this$_profileName : (0, _getDisplayName.default)((0, _nullthrows.default)(this._reactElement.type))})`,
      onRender: _trackReactProfilerRender.default
    }, this._reactElement), this);
  }

  disconnectedCallback() {
    if (this._reactElement == null) {
      return;
    }

    _reactDom.default.unmountComponentAtNode(this);
  }

}

let reactMountRootElement;

try {
  customElements.define('nuclide-react-mount-root', ReactMountRootElement);
  reactMountRootElement = ReactMountRootElement;
} catch (e) {
  // Element was already registered. Retrieve its constructor:
  const oldElem = document.createElement('nuclide-react-mount-root');
  (0, _assert.default)(oldElem.constructor.name === 'ReactMountRootElement');
  reactMountRootElement = oldElem.constructor;
}

var _default = reactMountRootElement;
exports.default = _default;
