// @atom-ide-community/nuclide-commons-ui/RegExpFilter.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _RegExpFilter = _interopRequireWildcard(require("./RegExpFilter"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class Example extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      text: '',
      isRegExp: false,
      invalid: false
    };

    this._handleChange = change => {
      const {
        invalid
      } = (0, _RegExpFilter.getFilterPattern)(change.text, change.isRegExp);
      this.setState({ ...change,
        invalid
      });
    };
  }

  render() {
    const {
      text,
      isRegExp,
      invalid
    } = this.state;
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_RegExpFilter.default, {
      value: {
        text,
        isRegExp,
        invalid
      },
      onChange: this._handleChange
    })));
  }

}

var _default = {
  sectionName: 'RegExp Filter',
  description: 'An input for filtering that allows the use of regular expressions.',
  examples: [{
    title: 'RegExpFilter',
    component: Example
  }]
};
exports.default = _default;
