// @atom-ide-community/nuclide-commons-ui/SettingsColorInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _SettingsUtils = require("./SettingsUtils");

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class SettingsColorInput extends React.Component {
  constructor(...args) {
    super(...args);

    this._handleChange = event => {
      const value = event.target.value;
      this.props.onChange(value);
    };
  }

  render() {
    const {
      keyPath,
      title,
      description,
      value
    } = this.props;
    const id = (0, _SettingsUtils.normalizeIdentifier)(keyPath);
    return /*#__PURE__*/React.createElement("div", {
      className: "color"
    }, /*#__PURE__*/React.createElement("label", {
      className: "control-label"
    }, /*#__PURE__*/React.createElement("input", {
      id: id,
      type: "color",
      onChange: this._handleChange,
      value: value.toHexString()
    }), /*#__PURE__*/React.createElement("div", {
      className: "setting-title"
    }, title)), /*#__PURE__*/React.createElement("div", {
      className: "setting-description"
    }, description));
  }

}

exports.default = SettingsColorInput;
