// @atom-ide-community/nuclide-commons-ui/showModal.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Button = require("./Button");

var _showModal = _interopRequireDefault(require("./showModal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
function ModalButton() {
  return /*#__PURE__*/React.createElement(_Button.Button, {
    onClick: showExampleModal
  }, "Show Modal");
}

function showExampleModal() {
  (0, _showModal.default)(({
    dismiss
  }) => {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, "I'm a modal. You can add any content you like. I have all the standard behavior, like obeying the \"core:cancel\" command!"), /*#__PURE__*/React.createElement(_Button.Button, {
      onClick: dismiss
    }, "Hide Modal"));
  });
}

const ModalExamples = {
  sectionName: 'Modal',
  description: 'Overlays that cover the entire screen. ',
  examples: [{
    title: 'Click the button to toggle a modal:',
    component: ModalButton
  }]
};
exports.ModalExamples = ModalExamples;
