// @atom-ide-community/nuclide-commons-ui/TabbableContainer.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabbableContainerExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _AtomInput = require("./AtomInput");

var _Block = require("./Block");

var _Button = require("./Button");

var _RadioGroup = _interopRequireDefault(require("./RadioGroup"));

var _TabbableContainer = _interopRequireDefault(require("./TabbableContainer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const labels = ['radio 1', 'radio 2', 'radio 3'];

class FormExample extends React.Component {
  constructor(props) {
    super(props);

    this.onSelectedChange = selectedIndex => {
      this.setState({
        selectedIndex
      });
    };

    this.state = {
      selectedIndex: 0
    };
  }

  render() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
      disabled: false,
      initialValue: "input field 1",
      placeholderText: "placeholder text"
    })), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
      disabled: false,
      initialValue: "input field 2",
      placeholderText: "placeholder text"
    })), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
      disabled: false,
      initialValue: "input field 3",
      placeholderText: "placeholder text"
    })), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_RadioGroup.default, {
      selectedIndex: this.state.selectedIndex,
      optionLabels: labels,
      onSelectedChange: this.onSelectedChange
    })), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
      disabled: false,
      initialValue: "input field 4",
      placeholderText: "placeholder text"
    })), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Button.Button, {
      className: "inline-block",
      size: "SMALL"
    }, "button 1"), /*#__PURE__*/React.createElement(_Button.Button, {
      className: "inline-block",
      size: "SMALL"
    }, "button 2"), /*#__PURE__*/React.createElement(_Button.Button, {
      className: "inline-block",
      size: "SMALL"
    }, "button 3")));
  }

}

const ContainedTabbableContainerExample = () => /*#__PURE__*/React.createElement(_TabbableContainer.default, {
  contained: true
}, /*#__PURE__*/React.createElement(FormExample, null));

const UncontainedTabbableContainerExample = () => /*#__PURE__*/React.createElement(_TabbableContainer.default, {
  contained: false
}, /*#__PURE__*/React.createElement(FormExample, null));

const TabbableContainerExamples = {
  sectionName: 'TabbableContainer',
  description: 'Allows tabbing and shift-tabbing to change the focus of the inputs.',
  examples: [{
    title: 'Contained (focus will be contained in this section)',
    component: ContainedTabbableContainerExample
  }, {
    title: 'Uncontained',
    component: UncontainedTabbableContainerExample
  }]
};
exports.TabbableContainerExamples = TabbableContainerExamples;
