// @atom-ide-community/nuclide-commons-ui/Tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Icon = require("./Icon");

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class Tabs extends React.Component {
  constructor(...args) {
    super(...args);

    this._handleTabChange = selectedTabName => {
      if (typeof this.props.onActiveTabChange === 'function') {
        this.props.onActiveTabChange((0, _nullthrows.default)(this.props.tabs.find(tab => tab.name === selectedTabName)));
      }
    };

    this._renderTabMenu = () => {
      const closeButton = this.props.closeable ? /*#__PURE__*/React.createElement("div", {
        className: "close-icon",
        onClick: this.props.onClose
      }) : null;
      const tabs = this.props.tabs.map(tab => {
        const icon = tab.icon == null ? null : /*#__PURE__*/React.createElement(_Icon.Icon, {
          icon: tab.icon
        });
        const handler = {};
        handler[this.props.triggeringEvent] = this._handleTabChange.bind(this, tab.name);
        return /*#__PURE__*/React.createElement("li", _extends({
          className: (0, _classnames.default)({
            tab: true,
            active: this.props.activeTabName === tab.name,
            growable: this.props.growable
          }),
          key: tab.name,
          title: tab.name
        }, handler), /*#__PURE__*/React.createElement("div", {
          className: "title"
        }, icon, tab.tabContent), closeButton);
      });
      return /*#__PURE__*/React.createElement("ul", {
        className: "tab-bar list-inline inset-panel"
      }, tabs);
    };
  }

  render() {
    return /*#__PURE__*/React.createElement("div", {
      className: "nuclide-tabs"
    }, this._renderTabMenu());
  }

}

exports.default = Tabs;
Tabs.defaultProps = {
  closeable: false,
  triggeringEvent: 'onClick',
  growable: false
};
