// @atom-ide-community/nuclide-commons-ui/Tabs.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Tabs = _interopRequireDefault(require("./Tabs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const tabs = [{
  name: 'one',
  tabContent: /*#__PURE__*/React.createElement("div", null, "One")
}, {
  name: 'two',
  tabContent: /*#__PURE__*/React.createElement("div", null, "Two")
}, {
  name: 'three',
  tabContent: /*#__PURE__*/React.createElement("div", null, "Three")
}, {
  name: 'four',
  tabContent: /*#__PURE__*/React.createElement("div", null, "Four")
}, {
  name: 'five',
  tabContent: /*#__PURE__*/React.createElement("div", null, "Five")
}];

class TabExample extends React.Component {
  constructor(props) {
    super(props);

    this.handleTabChange = newTabName => {
      this.setState({
        activeTabName: newTabName.name
      });
    };

    this.state = {
      activeTabName: 'one'
    };
  }

  render() {
    const {
      activeTabName
    } = this.state;
    return /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Tabs.default, {
      tabs: tabs,
      activeTabName: activeTabName,
      triggeringEvent: "onClick",
      onActiveTabChange: this.handleTabChange
    }), /*#__PURE__*/React.createElement("div", {
      style: {
        padding: '2em 0 2em 0'
      }
    }, "Showing content for tab \"", activeTabName, "\"."));
  }

}

const TabExamples = {
  sectionName: 'Tabs',
  description: '',
  examples: [{
    title: '',
    component: TabExample
  }]
};
exports.TabExamples = TabExamples;
