// @atom-ide-community/nuclide-commons-ui/TextInputs.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextInputExamples = void 0;

var _atom = require("atom");

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _AtomInput = require("./AtomInput");

var _AtomTextEditor = require("./AtomTextEditor");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const AtomInputExample = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  disabled: false,
  initialValue: "atom input",
  placeholderText: "placeholder text"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  disabled: true,
  initialValue: "disabled atom input",
  placeholderText: "placeholder text"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  initialValue: "xs atom input",
  placeholderText: "placeholder text",
  size: "xs"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  initialValue: "sm atom input",
  placeholderText: "placeholder text",
  size: "sm"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  initialValue: "lg atom input",
  placeholderText: "placeholder text",
  size: "lg"
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  initialValue: "unstyled atom input",
  placeholderText: "placeholder text",
  unstyled: true
})), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_AtomInput.AtomInput, {
  initialValue: "atom input with custom width",
  placeholderText: "placeholder text",
  width: 200
})));

const buffer1 = new _atom.TextBuffer({
  text: '/**\n * Hi!\n */\n\n// I am a TextBuffer.\nconst a = 42;'
});
const buffer2 = new _atom.TextBuffer({
  text: '/**\n * Hi!\n */\n\n// I am a read-only, gutter-less TextBuffer.\nconst a = 42;'
});
const editorWrapperStyle = {
  display: 'flex',
  flexGrow: 1,
  height: '12em',
  boxShadow: '0 0 20px 0 rgba(0, 0, 0, 0.3)'
};

const AtomTextEditorExample = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement("div", {
  style: editorWrapperStyle
}, /*#__PURE__*/React.createElement(_AtomTextEditor.AtomTextEditor, {
  gutterHidden: false,
  readOnly: false,
  syncTextContents: false,
  autoGrow: false,
  path: "aJavaScriptFile.js",
  textBuffer: buffer1
})), /*#__PURE__*/React.createElement("div", {
  style: { ...editorWrapperStyle,
    marginTop: '2em'
  }
}, /*#__PURE__*/React.createElement(_AtomTextEditor.AtomTextEditor, {
  gutterHidden: true,
  readOnly: true,
  syncTextContents: false,
  autoGrow: false,
  path: "aJavaScriptFile.js",
  textBuffer: buffer2
})));

const TextInputExamples = {
  sectionName: 'Text Inputs',
  description: '',
  examples: [{
    title: 'AtomInput',
    component: AtomInputExample
  }, {
    title: 'AtomTextEditor',
    component: AtomTextEditorExample
  }]
};
exports.TextInputExamples = TextInputExamples;
