// @atom-ide-community/nuclide-commons-ui/Toolbar.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolbarExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Toolbar = require("./Toolbar");

var _ToolbarCenter = require("./ToolbarCenter");

var _ToolbarLeft = require("./ToolbarLeft");

var _ToolbarRight = require("./ToolbarRight");

var _Button = require("./Button");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const ToolbarExampleLeft = () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Toolbar.Toolbar, {
  location: "top"
}, /*#__PURE__*/React.createElement(_ToolbarLeft.ToolbarLeft, null, /*#__PURE__*/React.createElement("div", null, "a toolbar can have multiple children,"), /*#__PURE__*/React.createElement(_Button.Button, null, "such as this button.")))), /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement("div", null, "Be sure to use ", '<ToolbarLeft/>, <ToolbarLeft/>, and <ToolbarLeft/>', " as children.")));

const ToolbarExampleCenter = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Toolbar.Toolbar, {
  location: "top"
}, /*#__PURE__*/React.createElement(_ToolbarCenter.ToolbarCenter, null, /*#__PURE__*/React.createElement("div", null, "Example of ", '<ToolbarCenter />', "."))));

const ToolbarExampleRight = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Toolbar.Toolbar, {
  location: "top"
}, /*#__PURE__*/React.createElement(_ToolbarRight.ToolbarRight, null, /*#__PURE__*/React.createElement("div", null, "Example of ", '<ToolbarRight />'))));

const ToolbarExampleMultiple = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Toolbar.Toolbar, {
  location: "top"
}, /*#__PURE__*/React.createElement(_ToolbarLeft.ToolbarLeft, null, /*#__PURE__*/React.createElement("div", null, "You can combine")), /*#__PURE__*/React.createElement(_ToolbarCenter.ToolbarCenter, null, /*#__PURE__*/React.createElement("div", null, "the various kinds")), /*#__PURE__*/React.createElement(_ToolbarRight.ToolbarRight, null, /*#__PURE__*/React.createElement("div", null, "of aligners."))));

const ToolbarExamples = {
  sectionName: 'Toolbar',
  description: '',
  examples: [{
    title: 'Left Toolbar',
    component: ToolbarExampleLeft
  }, {
    title: 'Center Toolbar',
    component: ToolbarExampleCenter
  }, {
    title: 'Right Toolbar',
    component: ToolbarExampleRight
  }, {
    title: 'Combining Toolbar aligners',
    component: ToolbarExampleMultiple
  }]
};
exports.ToolbarExamples = ToolbarExamples;
