// @atom-ide-community/nuclide-commons-ui/Tree.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Block = require("./Block");

var _Icon = require("./Icon");

var _Tree = require("./Tree");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
const BasicTreeExample = () => /*#__PURE__*/React.createElement("div", null, "Trees", /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Tree.TreeList, null, /*#__PURE__*/React.createElement(_Tree.TreeItem, null, "TreeItem 1"), /*#__PURE__*/React.createElement(_Tree.TreeItem, null, "TreeItem 2"), /*#__PURE__*/React.createElement(_Tree.NestedTreeItem, {
  title: /*#__PURE__*/React.createElement("span", null, "NestedTreeItem 1 -- click me!"),
  onSelect: handleSelect,
  onConfirm: handleConfirm,
  onTripleClick: handleTripleClick,
  selected: true
}, /*#__PURE__*/React.createElement(_Tree.TreeItem, null, "TreeItem 3"), /*#__PURE__*/React.createElement(_Tree.TreeItem, null, "TreeItem 4")), /*#__PURE__*/React.createElement(_Tree.NestedTreeItem, {
  title: /*#__PURE__*/React.createElement("span", null, "NestedTreeItem 2"),
  collapsed: true
}))));

const AtomStyleguideTreeExample = () => /*#__PURE__*/React.createElement(_Block.Block, null, /*#__PURE__*/React.createElement(_Tree.TreeList, {
  showArrows: true
}, /*#__PURE__*/React.createElement(_Tree.NestedTreeItem, {
  title: /*#__PURE__*/React.createElement(_Icon.Icon, {
    icon: "file-directory"
  }, "A Directory")
}, /*#__PURE__*/React.createElement(_Tree.NestedTreeItem, {
  collapsed: false,
  title: /*#__PURE__*/React.createElement(_Icon.Icon, {
    icon: "file-directory"
  }, "Nested Directory")
}, /*#__PURE__*/React.createElement(_Tree.TreeItem, null, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-text"
}, "File one"))), /*#__PURE__*/React.createElement(_Tree.NestedTreeItem, {
  collapsed: true,
  title: /*#__PURE__*/React.createElement(_Icon.Icon, {
    icon: "file-directory"
  }, "Collapsed Nested Directory")
}, /*#__PURE__*/React.createElement(_Tree.TreeItem, null, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-text"
}, "File one"))), /*#__PURE__*/React.createElement(_Tree.TreeItem, null, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-text"
}, "File one")), /*#__PURE__*/React.createElement(_Tree.TreeItem, {
  selected: true
}, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-text"
}, "File three .selected!"))), /*#__PURE__*/React.createElement(_Tree.TreeItem, null, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-text"
}, ".icon-file-text")), /*#__PURE__*/React.createElement(_Tree.TreeItem, null, /*#__PURE__*/React.createElement(_Icon.Icon, {
  icon: "file-symlink-file"
}, ".icon-file-symlink-file"))));

const TreeExamples = {
  sectionName: 'Trees',
  description: 'Expandable, hierarchical lists.',
  examples: [{
    title: 'Basic Tree',
    component: BasicTreeExample
  }, {
    title: 'Reproducing the Atom style guide example:',
    component: AtomStyleguideTreeExample
  }]
};
exports.TreeExamples = TreeExamples;

function handleSelect() {
  atom.notifications.addInfo('selected!');
}

function handleConfirm() {
  atom.notifications.addInfo('confirmed!');
}

function handleTripleClick() {
  atom.notifications.addInfo('triple clicked!');
}
