// @atom-ide-community/nuclide-commons-ui/UnstyledButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class UnstyledButton extends _react.default.Component {
  constructor(...args) {
    super(...args);
    this.props = void 0;
    this._node = void 0;

    this._setRef = node => this._node = node;
  }

  focus() {
    (0, _nullthrows.default)(this._node).focus();
  }

  render() {
    const {
      block,
      className,
      ...props
    } = this.props;
    const classes = (0, _classnames.default)('btn', 'nuclide-ui-unstyled-button', className, {
      'btn-block': block
    }); // eslint-disable-next-line nuclide-internal/use-nuclide-ui-components

    return /*#__PURE__*/_react.default.createElement("button", _extends({
      className: classes,
      ref: this._setRef
    }, props));
  }

}

exports.default = UnstyledButton;
