// atom-ide-base/commons-atom/ProviderRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProviderRegistry = void 0;
const atom_1 = require("atom");
class ProviderRegistry {
    constructor() {
        this.providers = [];
    }
    addProvider(provider) {
        const index = this.providers.findIndex((p) => provider.priority > p.priority);
        if (index === -1) {
            this.providers.push(provider);
        }
        else {
            this.providers.splice(index, 0, provider);
        }
        return new atom_1.Disposable(() => {
            this.removeProvider(provider);
        });
    }
    removeProvider(provider) {
        const index = this.providers.indexOf(provider);
        if (index !== -1) {
            this.providers.splice(index, 1);
        }
    }
    // TODO deprecate since there can be N providers.
    getProviderForEditor(editor) {
        const grammar = editor.getGrammar().scopeName;
        return this.findProvider(grammar);
    }
    // TODO create an ordering or priority aware util to prefer instead.
    getAllProvidersForEditor(editor) {
        const grammar = editor.getGrammar().scopeName;
        return this.findAllProviders(grammar);
    }
    findProvider(grammar) {
        for (const provider of this.findAllProviders(grammar)) {
            return provider;
        }
        return null;
    }
    /**
     * Iterates over all providers matching the grammar, in priority order.
     */
    *findAllProviders(grammar) {
        for (const provider of this.providers) {
            if (provider.grammarScopes == null || provider.grammarScopes.indexOf(grammar) !== -1) {
                yield provider;
            }
        }
    }
}
exports.ProviderRegistry = ProviderRegistry;
