// lemons/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Success = exports.Result = exports.Ok = exports.Nothing = exports.Maybe = exports.Loading = exports.LazyResult = exports.Just = exports.Initial = exports.Failure = exports.Err = undefined;

var _LazyResult = require('./LazyResult');

var _LazyResult2 = _interopRequireDefault(_LazyResult);

var _Maybe = require('./Maybe');

var _Maybe2 = _interopRequireDefault(_Maybe);

var _Result = require('./Result');

var _Result2 = _interopRequireDefault(_Result);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Err = _Result.Err;
exports.Failure = _LazyResult.Failure;
exports.Initial = _LazyResult.Initial;
exports.Just = _Maybe.Just;
exports.LazyResult = _LazyResult2.default;
exports.Loading = _LazyResult.Loading;
exports.Maybe = _Maybe2.default;
exports.Nothing = _Maybe.Nothing;
exports.Ok = _Result.Ok;
exports.Result = _Result2.default;
exports.Success = _LazyResult.Success; //  strict
