// @nteract/types/lib/core/hosts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeRemoteKernelRecord = exports.makeLocalKernelRecord = exports.makeDesktopHostRecord = exports.makeJupyterHostRecord = undefined;

var _immutable = require("immutable");

var makeJupyterHostRecord = exports.makeJupyterHostRecord = (0, _immutable.Record)({
  type: "jupyter",
  ref: null,
  id: null,
  selectedKernelRef: null,
  kernelspecsRef: null,
  defaultKernelName: null,
  activeKernelRefs: (0, _immutable.List)(),
  token: null,
  serverUrl: null,
  crossDomain: false
});
var makeDesktopHostRecord = exports.makeDesktopHostRecord = (0, _immutable.Record)({
  type: "local",
  ref: null,
  id: null,
  selectedKernelRef: null,
  kernelspecsRef: null,
  defaultKernelName: null,
  activeKernelRefs: (0, _immutable.List)()
});

var makeLocalKernelRecord = exports.makeLocalKernelRecord = (0, _immutable.Record)({
  type: "zeromq",
  ref: null,
  kernelSpecName: null,
  name: null,
  lastActivity: null,
  channels: null,
  status: null,
  spawn: null,
  connectionFile: null
});

var makeRemoteKernelRecord = exports.makeRemoteKernelRecord = (0, _immutable.Record)({
  type: "websocket",
  id: null,
  ref: null,
  kernelSpecName: null,
  name: null,
  lastActivity: null,
  channels: null,
  status: null
});
