// @nteract/types/lib/core/records.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommsRecord = exports.makeDocumentRecord = exports.makeAppRecord = exports.makeKernelspec = exports.makeCommunicationKernelspecs = exports.makeRemoteKernelRecord = exports.makeJupyterHostRecord = exports.makeDesktopHostRecord = exports.makeLocalKernelRecord = exports.createKernelspecsRef = exports.createKernelRef = exports.createHostRef = undefined;

var _refs = require("./refs");

Object.defineProperty(exports, "createHostRef", {
  enumerable: true,
  get: function get() {
    return _refs.createHostRef;
  }
});
Object.defineProperty(exports, "createKernelRef", {
  enumerable: true,
  get: function get() {
    return _refs.createKernelRef;
  }
});
Object.defineProperty(exports, "createKernelspecsRef", {
  enumerable: true,
  get: function get() {
    return _refs.createKernelspecsRef;
  }
});

var _hosts = require("./hosts");

Object.defineProperty(exports, "makeLocalKernelRecord", {
  enumerable: true,
  get: function get() {
    return _hosts.makeLocalKernelRecord;
  }
});
Object.defineProperty(exports, "makeDesktopHostRecord", {
  enumerable: true,
  get: function get() {
    return _hosts.makeDesktopHostRecord;
  }
});
Object.defineProperty(exports, "makeJupyterHostRecord", {
  enumerable: true,
  get: function get() {
    return _hosts.makeJupyterHostRecord;
  }
});
Object.defineProperty(exports, "makeRemoteKernelRecord", {
  enumerable: true,
  get: function get() {
    return _hosts.makeRemoteKernelRecord;
  }
});

var _kernelspecs = require("./kernelspecs");

Object.defineProperty(exports, "makeCommunicationKernelspecs", {
  enumerable: true,
  get: function get() {
    return _kernelspecs.makeCommunicationKernelspecs;
  }
});
Object.defineProperty(exports, "makeKernelspec", {
  enumerable: true,
  get: function get() {
    return _kernelspecs.makeKernelspec;
  }
});

var _immutable = require("immutable");

var version = require("../../package.json").version;
var makeAppRecord = exports.makeAppRecord = (0, _immutable.Record)({
  kernel: null,
  host: null,
  githubToken: null,
  notificationSystem: null,
  isSaving: false,
  lastSaved: null,
  configLastSaved: null,
  error: null,
  // set the default version to @nteract/core's version
  version: "@nteract/core@" + version
});

var makeDocumentRecord = exports.makeDocumentRecord = (0, _immutable.Record)({
  notebook: null,
  savedNotebook: null,
  // $FlowFixMe: Immutable
  transient: new _immutable.Map({
    keyPathsForDisplays: new _immutable.Map()
  }),
  cellPagers: new _immutable.Map(),
  stickyCells: new _immutable.Set(),
  editorFocused: null,
  cellFocused: null,
  copied: new _immutable.Map(),
  filename: ""
});
var CommsRecord = exports.CommsRecord = (0, _immutable.Record)({
  targets: new _immutable.Map(),
  info: new _immutable.Map(),
  models: new _immutable.Map()
});
