// @nteract/transform-vega/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const react_1 = require("./react");
exports.Vega = (mediaType) => {
    const embed = ({ data, options, onResult, onError }) => React.createElement(react_1.VegaEmbed, { mediaType: mediaType, spec: data, options: options, resultHandler: onResult, errorHandler: onError });
    embed.defaultProps = {
        mediaType
    };
    embed.MIMETYPE = mediaType;
    return embed;
};
exports.Vega2 = exports.Vega("application/vnd.vega.v2+json");
exports.Vega3 = exports.Vega("application/vnd.vega.v3+json");
exports.Vega4 = exports.Vega("application/vnd.vega.v4+json");
exports.Vega5 = exports.Vega("application/vnd.vega.v5+json");
exports.VegaLite1 = exports.Vega("application/vnd.vegalite.v1+json");
exports.VegaLite2 = exports.Vega("application/vnd.vegalite.v2+json");
exports.VegaLite3 = exports.Vega("application/vnd.vegalite.v3+json");
exports.VegaLite4 = exports.Vega("application/vnd.vegalite.v4+json");
