// atom-languageclient/build/lib/logger.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
let ConsoleLogger = exports.ConsoleLogger = class ConsoleLogger {

  constructor(prefix) {
    this.prefix = prefix;
  }

  warn(...args) {
    console.warn(...this.format(args));
  }

  error(...args) {
    console.error(...this.format(args));
  }

  info(...args) {
    console.info(...this.format(args));
  }

  debug(...args) {
    console.debug(...this.format(args));
  }

  log(...args) {
    console.log(...this.format(args));
  }

  format(args) {
    args = args.filter(a => a != null);
    if (typeof args[0] === 'string') {
      if (args.length === 1) {
        return [`${this.prefix} ${args[0]}`];
      } else if (args.length === 2) {
        return [`${this.prefix} ${args[0]}`, args[1]];
      } else {
        return [`${this.prefix} ${args[0]}`, args.slice(1)];
      }
    }

    return [`${this.prefix}`, args];
  }
};
let NullLogger = exports.NullLogger = class NullLogger {
  warn(args) {}
  error(args) {}
  info(args) {}
  log(args) {}
  debug(args) {}
};
