// atom-languageclient/build/lib/logger.js
"use strict";
/* eslint-disable no-console */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilteredLogger = exports.NullLogger = exports.ConsoleLogger = void 0;
class ConsoleLogger {
    constructor(prefix) {
        this.prefix = prefix;
    }
    warn(...args) {
        console.warn(...this.format(args));
    }
    error(...args) {
        console.error(...this.format(args));
    }
    info(...args) {
        console.info(...this.format(args));
    }
    debug(...args) {
        console.debug(...this.format(args));
    }
    log(...args) {
        console.log(...this.format(args));
    }
    format(args_) {
        const args = args_.filter((a) => a != null);
        if (typeof args[0] === "string") {
            if (args.length === 1) {
                return [`${this.prefix} ${args[0]}`];
            }
            else if (args.length === 2) {
                return [`${this.prefix} ${args[0]}`, args[1]];
            }
            else {
                return [`${this.prefix} ${args[0]}`, args.slice(1)];
            }
        }
        return [`${this.prefix}`, args];
    }
}
exports.ConsoleLogger = ConsoleLogger;
class NullLogger {
    warn(..._args) { }
    error(..._args) { }
    info(..._args) { }
    log(..._args) { }
    debug(..._args) { }
}
exports.NullLogger = NullLogger;
class FilteredLogger {
    constructor(logger, predicate) {
        this._logger = logger;
        this._predicate = predicate || ((_level, _args) => true);
    }
    warn(...args) {
        if (this._predicate("warn", args)) {
            this._logger.warn(...args);
        }
    }
    error(...args) {
        if (this._predicate("error", args)) {
            this._logger.error(...args);
        }
    }
    info(...args) {
        if (this._predicate("info", args)) {
            this._logger.info(...args);
        }
    }
    debug(...args) {
        if (this._predicate("debug", args)) {
            this._logger.debug(...args);
        }
    }
    log(...args) {
        if (this._predicate("log", args)) {
            this._logger.log(...args);
        }
    }
}
exports.FilteredLogger = FilteredLogger;
FilteredLogger.UserLevelFilter = (level, _args) => level === "warn" || level === "error";
FilteredLogger.DeveloperLevelFilter = (_level, _args) => true;
