// @atom-ide-community/nuclide-commons/AbortController.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.AbortSignal = void 0;

var _eventTargetShim = require("event-target-shim");

let _Symbol$toStringTag, _Symbol$toStringTag2;

_Symbol$toStringTag = Symbol.toStringTag;

class AbortSignal extends _eventTargetShim.EventTarget {
  constructor(...args) {
    super(...args);
    this.aborted = false;
    this.onabort = void 0;
  }

  // $FlowIssue: Computed properties are not supported
  get [_Symbol$toStringTag]() {
    return 'AbortSignal';
  }

}

exports.AbortSignal = AbortSignal;
(0, _eventTargetShim.defineEventAttribute)(AbortSignal.prototype, 'abort');
_Symbol$toStringTag2 = Symbol.toStringTag;

class AbortController {
  constructor() {
    this.signal = new AbortSignal();
  }

  abort() {
    // From whatwg spec, section 3.2:
    // If signal’s aborted flag is set, then return.
    if (this.signal.aborted) {
      return;
    } // Set signal’s aborted flag.


    this.signal.aborted = true; // Fire an event named abort at signal.
    // Note: event-target-shim converts objects to Events.

    this.signal.dispatchEvent({
      type: 'abort'
    });
  } // $FlowIssue: Computed properties are not supported


  get [_Symbol$toStringTag2]() {
    return 'AbortController';
  }

}

exports.default = AbortController;
