// @nteract/commutable/lib/cells.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeRawCell = exports.makeMarkdownCell = exports.makeCodeCell = void 0;
const immutable_1 = require("immutable");
exports.makeCodeCell = immutable_1.Record({
    cell_type: "code",
    execution_count: null,
    metadata: immutable_1.Map({
        collapsed: true,
        jupyter: immutable_1.Map({
            source_hidden: false,
            outputs_hidden: false,
        }),
        nteract: immutable_1.Map({
            transient: immutable_1.Map({
                deleting: false,
            }),
        }),
    }),
    source: "",
    outputs: immutable_1.List(),
});
exports.makeMarkdownCell = immutable_1.Record({
    cell_type: "markdown",
    metadata: immutable_1.Map({
        nteract: immutable_1.Map({
            transient: immutable_1.Map({
                deleting: false,
            }),
        }),
    }),
    source: "",
});
exports.makeRawCell = immutable_1.Record({
    cell_type: "raw",
    metadata: immutable_1.Map({
        nteract: immutable_1.Map({
            transient: immutable_1.Map({
                deleting: false,
            }),
        }),
    }),
    source: "",
});
