// @nteract/types/lib/entities/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeEntitiesRecord = void 0;
const Immutable = __importStar(require("immutable"));
const contents_1 = require("./contents");
const hosts_1 = require("./hosts");
const kernels_1 = require("./kernels");
const kernelspecs_1 = require("./kernelspecs");
const messages_1 = require("./messages");
const modals_1 = require("./modals");
const sidebar_1 = require("./sidebar");
const transforms_1 = require("./transforms");
const comms_1 = require("./comms");
const editors_1 = require("./editors");
__exportStar(require("./comms"), exports);
__exportStar(require("./contents"), exports);
__exportStar(require("./hosts"), exports);
__exportStar(require("./kernels"), exports);
__exportStar(require("./kernel-info"), exports);
__exportStar(require("./kernelspecs"), exports);
__exportStar(require("./messages"), exports);
__exportStar(require("./modals"), exports);
__exportStar(require("./transforms"), exports);
__exportStar(require("./sidebar"), exports);
__exportStar(require("./editors"), exports);
exports.makeEntitiesRecord = Immutable.Record({
    comms: comms_1.makeCommsRecord(),
    contents: contents_1.makeContentsRecord(),
    hosts: hosts_1.makeHostsRecord(),
    kernels: kernels_1.makeKernelsRecord(),
    kernelspecs: kernelspecs_1.makeKernelspecsRecord(),
    modals: modals_1.makeModalsRecord(),
    transforms: transforms_1.makeTransformsRecord(),
    messages: messages_1.makeMessagesRecord(),
    sidebar: sidebar_1.makeSidebarRecord(),
    editors: editors_1.makeEditorsRecord(),
});
