// @nteract/types/lib/entities/kernel-info.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeKernelInfoRecord = exports.makeHelpLinkRecord = void 0;
const Immutable = __importStar(require("immutable"));
exports.makeHelpLinkRecord = Immutable.Record({
    text: "",
    url: "#"
});
// NOTE: This is not exported as the "makeXRecord" for KernelInfo so that we
//       don't have nullables and can rely on the proper types. The developer is
//       required to pass a fully instantiated KernelInfo into
//       makeKernelInfoRecord below.
const kernelInfoRecordMaker = Immutable.Record({
    protocolVersion: "0.0.0",
    implementation: "",
    implementationVersion: "",
    banner: "",
    helpLinks: Immutable.List(),
    languageName: "",
    languageVersion: "",
    mimetype: "",
    fileExtension: "",
    pygmentsLexer: "",
    codemirrorMode: "text/plain",
    nbconvertExporter: ""
});
function makeKernelInfoRecord(kernelInfo) {
    return kernelInfoRecordMaker(kernelInfo);
}
exports.makeKernelInfoRecord = makeKernelInfoRecord;
