// @atom-ide-community/nuclide-commons-atom/ProviderRegistry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _UniversalDisposable = _interopRequireDefault(require("@atom-ide-community/nuclide-commons/UniversalDisposable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class ProviderRegistry {
  constructor() {
    this._providers = void 0;
    this._providers = [];
  }

  addProvider(provider) {
    const index = this._providers.findIndex(p => provider.priority > p.priority);

    if (index === -1) {
      this._providers.push(provider);
    } else {
      this._providers.splice(index, 0, provider);
    }

    return new _UniversalDisposable.default(() => {
      this.removeProvider(provider);
    });
  }

  removeProvider(provider) {
    const index = this._providers.indexOf(provider);

    if (index !== -1) {
      this._providers.splice(index, 1);
    }
  } // TODO deprecate since there can be N providers.


  getProviderForEditor(editor) {
    const grammar = editor.getGrammar().scopeName;
    return this.findProvider(grammar);
  } // TODO create an ordering or priority aware util to prefer instead.


  getAllProvidersForEditor(editor) {
    const grammar = editor.getGrammar().scopeName;
    return this.findAllProviders(grammar);
  }

  findProvider(grammar) {
    for (const provider of this.findAllProviders(grammar)) {
      return provider;
    }

    return null;
  }
  /**
   * Iterates over all providers matching the grammar, in priority order.
   */


  *findAllProviders(grammar) {
    for (const provider of this._providers) {
      if (provider.grammarScopes == null || provider.grammarScopes.indexOf(grammar) !== -1) {
        yield provider;
      }
    }
  }

}

exports.default = ProviderRegistry;
