// atom-languageclient/build/lib/adapters/command-execution-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CommandExecutionAdapter {
    static canAdapt(serverCapabilities) {
        return serverCapabilities.executeCommandProvider != null;
    }
    static registerCustomCallbackForCommand(command, callback) {
        this.commandsCustomCallbacks.set(command, callback);
    }
    /** Returns a {Promise} */
    static executeCommand(connection, command, commandArgs) {
        const executeCommandParams = CommandExecutionAdapter.createExecuteCommandParams(command, commandArgs);
        const commandCustomCallback = this.commandsCustomCallbacks.get(command);
        return commandCustomCallback !== undefined
            ? commandCustomCallback(executeCommandParams)
            : connection.executeCommand(executeCommandParams);
    }
    static createExecuteCommandParams(command, commandArgs) {
        return {
            command,
            arguments: commandArgs,
        };
    }
}
exports.default = CommandExecutionAdapter;
CommandExecutionAdapter.commandsCustomCallbacks = new Map();
