// atom-languageclient/build/lib/adapters/diagnostic-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.diagnosticTypeToLSSeverity = exports.atomIdeDiagnosticToLSDiagnostic = void 0;
const ls = require("../languageclient");
const convert_1 = require("../convert");
const linter_push_v2_adapter_1 = require("./linter-push-v2-adapter");
/** @deprecated Use Linter V2 service */
class IdeDiagnosticAdapter extends linter_push_v2_adapter_1.default {
    constructor() {
        super(...arguments);
        this._diagnosticCodes = new Map();
    }
    /**
     * Public: Capture the diagnostics sent from a langguage server, convert them to the Linter V2 format and forward them
     * on to any attached {V2IndieDelegate}s.
     *
     * @deprecated Use Linter V2 service
     * @param params The {PublishDiagnosticsParams} received from the language server that should be captured and
     *   forwarded on to any attached {V2IndieDelegate}s.
     */
    captureDiagnostics(params) {
        const path = convert_1.default.uriToPath(params.uri);
        const codeMap = new Map();
        const messages = params.diagnostics.map((d) => {
            const linterMessage = this.diagnosticToV2Message(path, d);
            codeMap.set(getCodeKey(linterMessage.location.position, d.message), d.code);
            return linterMessage;
        });
        this._diagnosticMap.set(path, messages);
        this._diagnosticCodes.set(path, codeMap);
        this._indies.forEach((i) => i.setMessages(path, messages));
    }
    /**
     * Public: get diagnostics for the given linter messages
     *
     * @deprecated Use Linter V2 service
     * @param linterMessages An array of linter {V2Message}
     * @param editor
     * @returns An array of LS {Diagnostic[]}
     */
    getLSDiagnosticsForIdeDiagnostics(diagnostics, editor) {
        return diagnostics.map((diagnostic) => this.getLSDiagnosticForIdeDiagnostic(diagnostic, editor));
    }
    /**
     * Public: Get the {Diagnostic} that is associated with the given {atomIde.Diagnostic}.
     *
     * @deprecated Use Linter V2 service
     * @param diagnostic The {atomIde.Diagnostic} object to fetch the {Diagnostic} for.
     * @param editor
     * @returns The associated {Diagnostic}.
     */
    getLSDiagnosticForIdeDiagnostic(diagnostic, editor) {
        // Retrieve the stored diagnostic code if it exists.
        // Until the Linter API provides a place to store the code,
        // there's no real way for the code actions API to give it back to us.
        const converted = atomIdeDiagnosticToLSDiagnostic(diagnostic);
        if (diagnostic.range != null && diagnostic.text != null) {
            const code = this.getDiagnosticCode(editor, diagnostic.range, diagnostic.text);
            if (code != null) {
                converted.code = code;
            }
        }
        return converted;
    }
    /**
     * Private: Get the recorded diagnostic code for a range/message. Diagnostic codes are tricky because there's no
     * suitable place in the Linter API for them. For now, we'll record the original code for each range/message
     * combination and retrieve it when needed (e.g. for passing back into code actions)
     */
    getDiagnosticCode(editor, range, text) {
        const path = editor.getPath();
        if (path != null) {
            const diagnosticCodes = this._diagnosticCodes.get(path);
            if (diagnosticCodes != null) {
                return diagnosticCodes.get(getCodeKey(range, text)) || null;
            }
        }
        return null;
    }
}
exports.default = IdeDiagnosticAdapter;
/** @deprecated Use Linter V2 service */
function atomIdeDiagnosticToLSDiagnostic(diagnostic) {
    // TODO: support diagnostic codes and codeDescriptions
    // TODO!: support data
    return {
        range: convert_1.default.atomRangeToLSRange(diagnostic.range),
        severity: diagnosticTypeToLSSeverity(diagnostic.type),
        source: diagnostic.providerName,
        message: diagnostic.text || "",
    };
}
exports.atomIdeDiagnosticToLSDiagnostic = atomIdeDiagnosticToLSDiagnostic;
/** @deprecated Use Linter V2 service */
function diagnosticTypeToLSSeverity(type) {
    switch (type) {
        case "Error":
            return ls.DiagnosticSeverity.Error;
        case "Warning":
            return ls.DiagnosticSeverity.Warning;
        case "Info":
            return ls.DiagnosticSeverity.Information;
        default:
            throw Error(`Unexpected diagnostic type ${type}`);
    }
}
exports.diagnosticTypeToLSSeverity = diagnosticTypeToLSSeverity;
function getCodeKey(range, text) {
    return [].concat(...range.serialize(), text).join(",");
}
