// atom-languageclient/build/lib/adapters/rename-adapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const convert_1 = require("../convert");
class RenameAdapter {
    static canAdapt(serverCapabilities) {
        return serverCapabilities.renameProvider === true;
    }
    static getRename(connection, editor, point, newName) {
        return __awaiter(this, void 0, void 0, function* () {
            const edit = yield connection.rename(RenameAdapter.createRenameParams(editor, point, newName));
            if (edit === null) {
                return null;
            }
            if (edit.documentChanges) {
                return RenameAdapter.convertDocumentChanges(edit.documentChanges);
            }
            else if (edit.changes) {
                return RenameAdapter.convertChanges(edit.changes);
            }
            else {
                return null;
            }
        });
    }
    static createRenameParams(editor, point, newName) {
        return {
            textDocument: convert_1.default.editorToTextDocumentIdentifier(editor),
            position: convert_1.default.pointToPosition(point),
            newName,
        };
    }
    static convertChanges(changes) {
        const result = new Map();
        Object.keys(changes).forEach((uri) => {
            result.set(convert_1.default.uriToPath(uri), convert_1.default.convertLsTextEdits(changes[uri]));
        });
        return result;
    }
    static convertDocumentChanges(documentChanges) {
        const result = new Map();
        documentChanges.forEach((documentEdit) => {
            result.set(convert_1.default.uriToPath(documentEdit.textDocument.uri), convert_1.default.convertLsTextEdits(documentEdit.edits));
        });
        return result;
    }
}
exports.default = RenameAdapter;
