// atom-languageclient/build/lib/adapters/show-document-adapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.showDocument = exports.attach = void 0;
const electron_1 = require("electron");
const atom_1 = require("atom");
const convert_1 = require("../convert");
/** Public: Adapts the window/showDocument command to Atom's text editors or external programs. */
const ShowDocumentAdapter = {
    /** {@inheritDoc attach} */
    attach,
    /** {@inheritDoc showDocument} */
    showDocument,
};
// for consistency with other adapters
exports.default = ShowDocumentAdapter;
/** Public: Attach to a {LanguageClientConnection} to recieve requests to show documents. */
function attach(connection) {
    connection.onShowDocument(showDocument);
}
exports.attach = attach;
/**
 * Public: show documents inside Atom text editor or in external programs
 *
 * @param params The {ShowDocumentParams} received from the language server indicating the document to be displayed as
 *   well as other metadata.
 * @returns {Promise<ShowDocumentResult>} With a `success: boolean` property specifying if the operation was sucessful
 *   {@inheritDoc ShowDocumentParams}
 */
function showDocument(params) {
    var _a, _b, _c, _d;
    return __awaiter(this, void 0, void 0, function* () {
        try {
            if (!params.external) {
                // open using atom.workspace
                const view = yield atom.workspace.open(convert_1.default.uriToPath(params.uri), {
                    activateItem: params.takeFocus,
                    activatePane: params.takeFocus,
                    pending: true,
                    initialLine: (_b = (_a = params.selection) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _b !== void 0 ? _b : 0,
                    initialColumn: (_d = (_c = params.selection) === null || _c === void 0 ? void 0 : _c.start.character) !== null && _d !== void 0 ? _d : 0,
                });
                if (!view) {
                    return { success: false };
                }
                if (view instanceof atom_1.TextEditor && params.selection !== undefined) {
                    view.selectToBufferPosition(convert_1.default.positionToPoint(params.selection.end));
                }
            }
            else {
                // open using Electron
                electron_1.shell.openExternal(params.uri, { activate: params.takeFocus });
            }
            return { success: true };
        }
        catch (e) {
            atom.notifications.addError(e);
            return { success: false };
        }
    });
}
exports.showDocument = showDocument;
