// atom-languageclient/build/lib/adapters/datatip-adapter.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _languageclient = require('../languageclient');

var _convert = require('../convert');

var _convert2 = _interopRequireDefault(_convert);

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

// Public: Adapts the language server protocol "textDocument/hover" to the
// Atom IDE UI Datatip package.
let DatatipAdapter = class DatatipAdapter {

  // Public: Determine whether this adapter can be used to adapt a language server
  // based on the serverCapabilities matrix containing a hoverProvider.
  //
  // * `serverCapabilities` The {ServerCapabilities} of the language server to consider.
  //
  // Returns a {Boolean} indicating adapter can adapt the server based on the
  // given serverCapabilities.
  static canAdapt(serverCapabilities) {
    return serverCapabilities.hoverProvider === true;
  }

  // Public: Get the Datatip for this {Point} in a {TextEditor} by querying
  // the language server.
  //
  // * `connection` A {LanguageClientConnection} to the language server that will be queried
  //                for the hover text/datatip.
  // * `editor` The Atom {TextEditor} containing the text the Datatip should relate to.
  // * `point` The Atom {Point} containing the point within the text the Datatip should relate to.
  //
  // Returns a {Promise} containing the {Datatip} to display or {null} if no Datatip is available.
  getDatatip(connection, editor, point) {
    return _asyncToGenerator(function* () {
      const documentPositionParams = _convert2.default.editorToTextDocumentPositionParams(editor, point);

      const hover = yield connection.hover(documentPositionParams);
      if (hover == null || hover.contents == null ||
      // This intentionally covers both empty strings and empty arrays.
      hover.contents.length === 0) {
        return null;
      }

      const range = hover.range == null ? _utils2.default.getWordAtPosition(editor, point) : _convert2.default.lsRangeToAtomRange(hover.range);

      const markedStrings = (Array.isArray(hover.contents) ? hover.contents : [hover.contents]).map(function (str) {
        return DatatipAdapter.convertMarkedString(editor, str);
      });

      return { range, markedStrings };
    })();
  }

  static convertMarkedString(editor, markedString) {
    if (typeof markedString === 'object') {
      return {
        type: 'snippet',
        // TODO: find a better mapping from language -> grammar
        grammar: atom.grammars.grammarForScopeName(`source.${markedString.language}`) || editor.getGrammar(),
        value: markedString.value
      };
    }
    return { type: 'markdown', value: markedString };
  }
};
exports.default = DatatipAdapter;
module.exports = exports['default'];
