// atom-languageclient/build/lib/adapters/notifications-adapter.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _languageclient = require('../languageclient');

// Public: Adapts Atom's user notifications to those of the language server protocol.
let NotificationsAdapter = class NotificationsAdapter {
  // Public: Attach to a {LanguageClientConnection} to recieve events indicating
  // when user notifications should be displayed.
  static attach(connection, name) {
    connection.onShowMessage(m => NotificationsAdapter.onShowMessage(m, name));
    connection.onShowMessageRequest(m => NotificationsAdapter.onShowMessageRequest(m, name));
  }

  static onShowMessageRequest(params, name) {
    return new Promise((resolve, reject) => {
      let notification;
      const options = { dismissable: true, detail: name };
      if (params.actions) {
        options.buttons = params.actions.map(a => ({
          text: a.title,
          onDidClick: () => {
            resolve(a);
            if (notification != null) {
              notification.dismiss();
            }
          }
        }));
      }

      notification = addNotificationForMessage(params.type, params.message, {
        dismissable: true,
        detail: name
      });

      if (notification != null) {
        notification.onDidDismiss(() => {
          resolve(null);
        });
      }
    });
  }

  // Public: Show a notification message using the Atom notifications API.
  //
  // * `params` The {ShowMessageParams} received from the language server
  //            indicating the details of the notification to be displayed.
  // * `name`   The name of the language server so the user can identify the
  //            context of the message.
  static onShowMessage(params, name) {
    addNotificationForMessage(params.type, params.message, {
      dismissable: true,
      detail: name
    });
  }

  // Public: Convert a {MessageActionItem} from the language server into an
  // equivalent {NotificationButton} within Atom.
  //
  // * `actionItem` The {MessageActionItem} to be converted.
  //
  // Returns a {NotificationButton} equivalent to the {MessageActionItem} given.
  static actionItemToNotificationButton(actionItem) {
    return {
      text: actionItem.title
    };
  }
};
exports.default = NotificationsAdapter;


function addNotificationForMessage(messageType, message, options) {
  switch (messageType) {
    case _languageclient.MessageType.Error:
      return atom.notifications.addError(message, options);
    case _languageclient.MessageType.Warning:
      return atom.notifications.addWarning(message, options);
    case _languageclient.MessageType.Log:
      // console.log(params.message);
      return;
    case _languageclient.MessageType.Info:
    default:
      return atom.notifications.addInfo(message, options);
  }
}
module.exports = exports['default'];
