// atom-languageclient/build/lib/adapters/signature-help-adapter.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _atom = require('atom');

var _languageclient = require('../languageclient');

var _convert = require('../convert');

var _convert2 = _interopRequireDefault(_convert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SignatureHelpAdapter {

  constructor(server, grammarScopes) {
    this._disposables = new _atom.CompositeDisposable();

    this._connection = server.connection;
    this._capabilities = server.capabilities;
    this._grammarScopes = grammarScopes;
  }

  // Returns a {Boolean} indicating this adapter can adapt the server based on the
  // given serverCapabilities.
  static canAdapt(serverCapabilities) {
    return serverCapabilities.signatureHelpProvider != null;
  }

  dispose() {
    this._disposables.dispose();
  }

  attach(register) {
    const { signatureHelpProvider } = this._capabilities;
    (0, _assert2.default)(signatureHelpProvider != null);

    let triggerCharacters = null;
    if (Array.isArray(signatureHelpProvider.triggerCharacters)) {
      triggerCharacters = new Set(signatureHelpProvider.triggerCharacters);
    }

    this._disposables.add(register({
      priority: 1,
      grammarScopes: this._grammarScopes,
      triggerCharacters,
      getSignatureHelp: this.getSignatureHelp.bind(this)
    }));
  }

  // Public: Retrieves signature help for a given editor and position.
  getSignatureHelp(editor, point) {
    return this._connection.signatureHelp(_convert2.default.editorToTextDocumentPositionParams(editor, point));
  }
}
exports.default = SignatureHelpAdapter;
module.exports = exports['default'];
