// atom-languageclient/build/test/helpers.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createSpyConnection = createSpyConnection;
exports.createFakeEditor = createFakeEditor;

var _sinon = require('sinon');

var _sinon2 = _interopRequireDefault(_sinon);

var _atom = require('atom');

var _vscodeJsonrpc = require('vscode-jsonrpc');

var rpc = _interopRequireWildcard(_vscodeJsonrpc);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createSpyConnection() {
  return {
    listen: _sinon2.default.spy(),
    onClose: _sinon2.default.spy(),
    onError: _sinon2.default.spy(),
    onUnhandledNotification: _sinon2.default.spy(),
    onNotification: _sinon2.default.spy(),
    dispose: _sinon2.default.spy(),
    sendRequest: _sinon2.default.spy()
  };
}

function createFakeEditor(path) {
  const editor = new _atom.TextEditor();
  _sinon2.default.stub(editor, 'getSelectedBufferRange');
  _sinon2.default.spy(editor, 'setTextInBufferRange');
  editor.setTabLength(4);
  editor.setSoftTabs(true);
  editor.getBuffer().setPath(path || '/a/b/c/d.js');
  return editor;
}
