// atom-languageclient/build/test/languageclient.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ls = require("../lib/languageclient");
const sinon = require("sinon");
const chai_1 = require("chai");
const helpers_js_1 = require("./helpers.js");
const logger_1 = require("../lib/logger");
describe('LanguageClientConnection', () => {
    beforeEach(() => {
        global.sinon = sinon.sandbox.create();
    });
    afterEach(() => {
        global.sinon.restore();
    });
    it('listens to the RPC connection it is given', () => {
        const rpc = helpers_js_1.createSpyConnection();
        new ls.LanguageClientConnection(rpc, new logger_1.NullLogger());
        chai_1.expect(rpc.listen.called).equals(true);
    });
    it('disposes of the connection when it is disposed', () => {
        const rpc = helpers_js_1.createSpyConnection();
        const lc = new ls.LanguageClientConnection(rpc, new logger_1.NullLogger());
        chai_1.expect(rpc.dispose.called).equals(false);
        lc.dispose();
        chai_1.expect(rpc.dispose.called).equals(true);
    });
    describe('send requests', () => {
        const textDocumentPositionParams = {
            textDocument: { uri: 'file:///1/z80.asm' },
            position: { line: 24, character: 32 },
        };
        let lc;
        beforeEach(() => {
            lc = new ls.LanguageClientConnection(helpers_js_1.createSpyConnection(), new logger_1.NullLogger());
            sinon.spy(lc, '_sendRequest');
        });
        it('sends a request for initialize', () => __awaiter(this, void 0, void 0, function* () {
            const params = { capabilities: {} };
            yield lc.initialize(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('initialize');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for shutdown', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.shutdown();
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('shutdown');
        }));
        it('sends a request for completion', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.completion(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/completion');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for completionItemResolve', () => __awaiter(this, void 0, void 0, function* () {
            const completionItem = { label: 'abc' };
            yield lc.completionItemResolve(completionItem);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('completionItem/resolve');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(completionItem);
        }));
        it('sends a request for hover', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.hover(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/hover');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for signatureHelp', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.signatureHelp(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/signatureHelp');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for gotoDefinition', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.gotoDefinition(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/definition');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for findReferences', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.findReferences(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/references');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for documentHighlight', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.documentHighlight(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/documentHighlight');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for documentSymbol', () => __awaiter(this, void 0, void 0, function* () {
            yield lc.documentSymbol(textDocumentPositionParams);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/documentSymbol');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(textDocumentPositionParams);
        }));
        it('sends a request for workspaceSymbol', () => __awaiter(this, void 0, void 0, function* () {
            const params = { query: 'something' };
            yield lc.workspaceSymbol(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('workspace/symbol');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for codeAction', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
                range: {
                    start: { line: 1, character: 1 },
                    end: { line: 24, character: 32 },
                },
                context: { diagnostics: [] },
            };
            yield lc.codeAction(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/codeAction');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for codeLens', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
            };
            yield lc.codeLens(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/codeLens');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for codeLensResolve', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                range: {
                    start: { line: 1, character: 1 },
                    end: { line: 24, character: 32 },
                },
            };
            yield lc.codeLensResolve(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('codeLens/resolve');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for documentLink', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
            };
            yield lc.documentLink(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/documentLink');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for documentLinkResolve', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                range: {
                    start: { line: 1, character: 1 },
                    end: { line: 24, character: 32 },
                },
                target: 'abc.def.ghi',
            };
            yield lc.documentLinkResolve(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('documentLink/resolve');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for documentFormatting', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
                options: { tabSize: 6, insertSpaces: true, someValue: 'optional' },
            };
            yield lc.documentFormatting(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/formatting');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for documentRangeFormatting', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
                range: {
                    start: { line: 1, character: 1 },
                    end: { line: 24, character: 32 },
                },
                options: { tabSize: 6, insertSpaces: true, someValue: 'optional' },
            };
            yield lc.documentRangeFormatting(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/rangeFormatting');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for documentOnTypeFormatting', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: textDocumentPositionParams.textDocument,
                position: { line: 1, character: 1 },
                ch: '}',
                options: { tabSize: 6, insertSpaces: true, someValue: 'optional' },
            };
            yield lc.documentOnTypeFormatting(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/onTypeFormatting');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
        it('sends a request for rename', () => __awaiter(this, void 0, void 0, function* () {
            const params = {
                textDocument: { uri: 'file:///a/b.txt' },
                position: { line: 1, character: 2 },
                newName: 'abstractConstructorFactory',
            };
            yield lc.rename(params);
            chai_1.expect(lc._sendRequest.called).equals(true);
            chai_1.expect(lc._sendRequest.getCall(0).args[0]).equals('textDocument/rename');
            chai_1.expect(lc._sendRequest.getCall(0).args[1]).equals(params);
        }));
    });
    describe('send notifications', () => {
        const textDocumentItem = {
            uri: 'file:///best/bits.js',
            languageId: 'javascript',
            text: 'function a() { return "b"; };',
            version: 1,
        };
        const versionedTextDocumentIdentifier = {
            uri: 'file:///best/bits.js',
            version: 1,
        };
        let lc;
        beforeEach(() => {
            lc = new ls.LanguageClientConnection(helpers_js_1.createSpyConnection(), new logger_1.NullLogger());
            sinon.stub(lc, '_sendNotification');
        });
        it('exit sends notification', () => {
            lc.exit();
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('exit');
            chai_1.expect(lc._sendNotification.getCall(0).args.length).equals(1);
        });
        it('initialized sends notification', () => {
            lc.initialized();
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('initialized');
            const expected = {};
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).to.deep.equal(expected);
        });
        it('didChangeConfiguration sends notification', () => {
            const params = {
                settings: { a: { b: 'c' } },
            };
            lc.didChangeConfiguration(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('workspace/didChangeConfiguration');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
        it('didOpenTextDocument sends notification', () => {
            const params = {
                textDocument: textDocumentItem,
            };
            lc.didOpenTextDocument(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('textDocument/didOpen');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
        it('didChangeTextDocument sends notification', () => {
            const params = {
                textDocument: versionedTextDocumentIdentifier,
                contentChanges: [],
            };
            lc.didChangeTextDocument(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('textDocument/didChange');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
        it('didCloseTextDocument sends notification', () => {
            const params = {
                textDocument: textDocumentItem,
            };
            lc.didCloseTextDocument(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('textDocument/didClose');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
        it('didSaveTextDocument sends notification', () => {
            const params = {
                textDocument: textDocumentItem,
            };
            lc.didSaveTextDocument(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('textDocument/didSave');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
        it('didChangeWatchedFiles sends notification', () => {
            const params = { changes: [] };
            lc.didChangeWatchedFiles(params);
            chai_1.expect(lc._sendNotification.called).equals(true);
            chai_1.expect(lc._sendNotification.getCall(0).args[0]).equals('workspace/didChangeWatchedFiles');
            chai_1.expect(lc._sendNotification.getCall(0).args[1]).equals(params);
        });
    });
    describe('notification methods', () => {
        let lc;
        const eventMap = {};
        beforeEach(() => {
            lc = new ls.LanguageClientConnection(helpers_js_1.createSpyConnection(), new logger_1.NullLogger());
            sinon.stub(lc, '_onNotification').callsFake((message, callback) => {
                eventMap[message.method] = callback;
            });
        });
        it('onShowMessage calls back on window/showMessage', () => {
            let called = false;
            lc.onShowMessage(() => {
                called = true;
            });
            eventMap['window/showMessage']();
            chai_1.expect(called).equals(true);
        });
    });
});
