// @nteract/types/lib/entities/contents/notebook.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeNotebookContentRecord = exports.makeDocumentRecord = void 0;
// Vendor modules
const commutable_1 = require("@nteract/commutable");
const Immutable = __importStar(require("immutable"));
exports.makeDocumentRecord = Immutable.Record({
    type: "notebook",
    notebook: commutable_1.emptyNotebook,
    savedNotebook: commutable_1.emptyNotebook,
    transient: Immutable.Map({
        keyPathsForDisplays: Immutable.Map()
    }),
    cellPagers: Immutable.Map(),
    cellPrompts: Immutable.Map(),
    editorFocused: null,
    cellFocused: null,
    copied: null,
    kernelRef: null
});
exports.makeNotebookContentRecord = Immutable.Record({
    mimetype: null,
    created: null,
    format: "json",
    lastSaved: null,
    model: exports.makeDocumentRecord(),
    filepath: "",
    type: "notebook",
    writable: true,
    saving: false,
    loading: false,
    error: null,
    showHeaderEditor: false
});
