// @atom-ide-community/nuclide-commons-ui/VendorLib/atom-tabs/lib/tab-view.js
"use strict";

// Generated by CoffeeScript 1.12.2
(function () {
  var CompositeDisposable,
      Disposable,
      FileIcons,
      TabView,
      layout,
      path,
      ref,
      slice = [].slice;
  path = require('path');
  ref = require('atom'), Disposable = ref.Disposable, CompositeDisposable = ref.CompositeDisposable;
  FileIcons = require('./file-icons');
  layout = require('./layout');

  module.exports = TabView = function () {
    function TabView(arg) {
      var base, closeIcon, didClickCloseIcon, location;
      this.item = arg.item, this.pane = arg.pane, didClickCloseIcon = arg.didClickCloseIcon, this.tabs = arg.tabs, location = arg.location;

      if (typeof this.item.getPath === 'function') {
        this.path = this.item.getPath();
      }

      this.element = document.createElement('li');
      this.element.setAttribute('is', 'tabs-tab');

      if (['TextEditor', 'TestView'].indexOf(this.item.constructor.name) > -1) {
        this.element.classList.add('texteditor');
      }

      this.element.classList.add('tab', 'sortable');
      this.itemTitle = document.createElement('div');
      this.itemTitle.classList.add('title');
      this.element.appendChild(this.itemTitle);

      if (location === 'center' || !(typeof (base = this.item).isPermanentDockItem === "function" ? base.isPermanentDockItem() : void 0)) {
        closeIcon = document.createElement('div');
        closeIcon.classList.add('close-icon');
        closeIcon.onclick = didClickCloseIcon;
        this.element.appendChild(closeIcon);
      }

      this.subscriptions = new CompositeDisposable();
      this.handleEvents();
      this.updateDataAttributes();
      this.updateTitle();
      this.updateIcon();
      this.updateModifiedStatus();
      this.setupTooltip();
      layout.activate();

      if (this.isItemPending()) {
        this.itemTitle.classList.add('temp');
        this.element.classList.add('pending-tab');
      }

      this.element.ondrag = function (e) {
        return layout.drag(e);
      };

      this.element.ondragend = function (e) {
        return layout.end(e);
      };

      this.element.pane = this.pane;
      this.element.item = this.item;
      this.element.itemTitle = this.itemTitle;
      this.element.path = this.path;
    }

    TabView.prototype.handleEvents = function () {
      var base, iconChangedHandler, modifiedHandler, onDidChangeIconDisposable, onDidChangeModifiedDisposable, onDidChangePathDisposable, onDidChangeTitleDisposable, onDidSaveDisposable, onDidTerminatePendingStateDisposable, pathChangedHandler, titleChangedHandler;

      titleChangedHandler = function (_this) {
        return function () {
          return _this.updateTitle();
        };
      }(this);

      this.subscriptions.add(this.pane.onDidDestroy(function (_this) {
        return function () {
          return _this.destroy();
        };
      }(this)));

      if (typeof this.pane.onItemDidTerminatePendingState === 'function') {
        this.subscriptions.add(this.pane.onItemDidTerminatePendingState(function (_this) {
          return function (item) {
            if (item === _this.item) {
              return _this.clearPending();
            }
          };
        }(this)));
      } else if (typeof this.item.onDidTerminatePendingState === 'function') {
        onDidTerminatePendingStateDisposable = this.item.onDidTerminatePendingState(function (_this) {
          return function () {
            return _this.clearPending();
          };
        }(this));

        if (Disposable.isDisposable(onDidTerminatePendingStateDisposable)) {
          this.subscriptions.add(onDidTerminatePendingStateDisposable);
        } else {
          console.warn("::onDidTerminatePendingState does not return a valid Disposable!", this.item);
        }
      }

      if (typeof this.item.onDidChangeTitle === 'function') {
        onDidChangeTitleDisposable = this.item.onDidChangeTitle(titleChangedHandler);

        if (Disposable.isDisposable(onDidChangeTitleDisposable)) {
          this.subscriptions.add(onDidChangeTitleDisposable);
        } else {
          console.warn("::onDidChangeTitle does not return a valid Disposable!", this.item);
        }
      } else if (typeof this.item.on === 'function') {
        this.item.on('title-changed', titleChangedHandler);
        this.subscriptions.add({
          dispose: function (_this) {
            return function () {
              var base;
              return typeof (base = _this.item).off === "function" ? base.off('title-changed', titleChangedHandler) : void 0;
            };
          }(this)
        });
      }

      pathChangedHandler = function (_this) {
        return function (path1) {
          _this.path = path1;

          _this.updateDataAttributes();

          _this.updateTitle();

          return _this.updateTooltip();
        };
      }(this);

      if (typeof this.item.onDidChangePath === 'function') {
        onDidChangePathDisposable = this.item.onDidChangePath(pathChangedHandler);

        if (Disposable.isDisposable(onDidChangePathDisposable)) {
          this.subscriptions.add(onDidChangePathDisposable);
        } else {
          console.warn("::onDidChangePath does not return a valid Disposable!", this.item);
        }
      } else if (typeof this.item.on === 'function') {
        this.item.on('path-changed', pathChangedHandler);
        this.subscriptions.add({
          dispose: function (_this) {
            return function () {
              var base;
              return typeof (base = _this.item).off === "function" ? base.off('path-changed', pathChangedHandler) : void 0;
            };
          }(this)
        });
      }

      iconChangedHandler = function (_this) {
        return function () {
          return _this.updateIcon();
        };
      }(this);

      if (typeof this.item.onDidChangeIcon === 'function') {
        onDidChangeIconDisposable = typeof (base = this.item).onDidChangeIcon === "function" ? base.onDidChangeIcon(function (_this) {
          return function () {
            return _this.updateIcon();
          };
        }(this)) : void 0;

        if (Disposable.isDisposable(onDidChangeIconDisposable)) {
          this.subscriptions.add(onDidChangeIconDisposable);
        } else {
          console.warn("::onDidChangeIcon does not return a valid Disposable!", this.item);
        }
      } else if (typeof this.item.on === 'function') {
        this.item.on('icon-changed', iconChangedHandler);
        this.subscriptions.add({
          dispose: function (_this) {
            return function () {
              var base1;
              return typeof (base1 = _this.item).off === "function" ? base1.off('icon-changed', iconChangedHandler) : void 0;
            };
          }(this)
        });
      }

      modifiedHandler = function (_this) {
        return function () {
          return _this.updateModifiedStatus();
        };
      }(this);

      if (typeof this.item.onDidChangeModified === 'function') {
        onDidChangeModifiedDisposable = this.item.onDidChangeModified(modifiedHandler);

        if (Disposable.isDisposable(onDidChangeModifiedDisposable)) {
          this.subscriptions.add(onDidChangeModifiedDisposable);
        } else {
          console.warn("::onDidChangeModified does not return a valid Disposable!", this.item);
        }
      } else if (typeof this.item.on === 'function') {
        this.item.on('modified-status-changed', modifiedHandler);
        this.subscriptions.add({
          dispose: function (_this) {
            return function () {
              var base1;
              return typeof (base1 = _this.item).off === "function" ? base1.off('modified-status-changed', modifiedHandler) : void 0;
            };
          }(this)
        });
      }

      if (typeof this.item.onDidSave === 'function') {
        onDidSaveDisposable = this.item.onDidSave(function (_this) {
          return function (event) {
            _this.terminatePendingState();

            if (event.path !== _this.path) {
              _this.path = event.path;

              if (atom.config.get('tabs.enableVcsColoring')) {
                return _this.setupVcsStatus();
              }
            }
          };
        }(this));

        if (Disposable.isDisposable(onDidSaveDisposable)) {
          this.subscriptions.add(onDidSaveDisposable);
        } else {
          console.warn("::onDidSave does not return a valid Disposable!", this.item);
        }
      }

      this.subscriptions.add(atom.config.observe('tabs.showIcons', function (_this) {
        return function () {
          return _this.updateIconVisibility();
        };
      }(this)));
      return this.subscriptions.add(atom.config.observe('tabs.enableVcsColoring', function (_this) {
        return function (isEnabled) {
          if (isEnabled && _this.path != null) {
            return _this.setupVcsStatus();
          } else {
            return _this.unsetVcsStatus();
          }
        };
      }(this)));
    };

    TabView.prototype.setupTooltip = function () {
      var onMouseEnter;

      onMouseEnter = function (_this) {
        return function () {
          _this.mouseEnterSubscription.dispose();

          _this.hasBeenMousedOver = true;

          _this.updateTooltip();

          return _this.element.dispatchEvent(new CustomEvent('mouseenter', {
            bubbles: true
          }));
        };
      }(this);

      this.mouseEnterSubscription = {
        dispose: function (_this) {
          return function () {
            _this.element.removeEventListener('mouseenter', onMouseEnter);

            return _this.mouseEnterSubscription = null;
          };
        }(this)
      };
      return this.element.addEventListener('mouseenter', onMouseEnter);
    };

    TabView.prototype.updateTooltip = function () {
      if (!this.hasBeenMousedOver) {
        return;
      }

      this.destroyTooltip();

      if (this.path) {
        return this.tooltip = atom.tooltips.add(this.element, {
          title: this.path,
          html: false,
          delay: {
            show: 1000,
            hide: 100
          },
          placement: 'bottom'
        });
      }
    };

    TabView.prototype.destroyTooltip = function () {
      var ref1;

      if (!this.hasBeenMousedOver) {
        return;
      }

      return (ref1 = this.tooltip) != null ? ref1.dispose() : void 0;
    };

    TabView.prototype.destroy = function () {
      var ref1, ref2, ref3;

      if ((ref1 = this.subscriptions) != null) {
        ref1.dispose();
      }

      if ((ref2 = this.mouseEnterSubscription) != null) {
        ref2.dispose();
      }

      if ((ref3 = this.repoSubscriptions) != null) {
        ref3.dispose();
      }

      this.destroyTooltip();
      return this.element.remove();
    };

    TabView.prototype.updateDataAttributes = function () {
      var itemClass, ref1;

      if (this.path) {
        this.itemTitle.dataset.name = path.basename(this.path);
        this.itemTitle.dataset.path = this.path;
      } else {
        delete this.itemTitle.dataset.name;
        delete this.itemTitle.dataset.path;
      }

      if (itemClass = (ref1 = this.item.constructor) != null ? ref1.name : void 0) {
        return this.element.dataset.type = itemClass;
      } else {
        return delete this.element.dataset.type;
      }
    };

    TabView.prototype.updateTitle = function (arg) {
      var base, base1, i, len, ref1, ref2, ref3, ref4, tab, title, updateSiblings, useLongTitle;
      ref1 = arg != null ? arg : {}, updateSiblings = ref1.updateSiblings, useLongTitle = ref1.useLongTitle;

      if (this.updatingTitle) {
        return;
      }

      this.updatingTitle = true;

      if (updateSiblings === false) {
        title = this.item.getTitle();

        if (useLongTitle) {
          title = (ref2 = typeof (base = this.item).getLongTitle === "function" ? base.getLongTitle() : void 0) != null ? ref2 : title;
        }

        this.itemTitle.textContent = title;
      } else {
        title = this.item.getTitle();
        useLongTitle = false;
        ref3 = this.tabs;

        for (i = 0, len = ref3.length; i < len; i++) {
          tab = ref3[i];

          if (tab !== this) {
            if (tab.item.getTitle() === title) {
              tab.updateTitle({
                updateSiblings: false,
                useLongTitle: true
              });
              useLongTitle = true;
            }
          }
        }

        if (useLongTitle) {
          title = (ref4 = typeof (base1 = this.item).getLongTitle === "function" ? base1.getLongTitle() : void 0) != null ? ref4 : title;
        }

        this.itemTitle.textContent = title;
      }

      return this.updatingTitle = false;
    };

    TabView.prototype.updateIcon = function () {
      var base, names, ref1, ref2;

      if (this.iconName) {
        names = !Array.isArray(this.iconName) ? this.iconName.split(/\s+/g) : this.iconName;
        (ref1 = this.itemTitle.classList).remove.apply(ref1, ['icon', "icon-" + names[0]].concat(slice.call(names)));
      }

      if (this.iconName = typeof (base = this.item).getIconName === "function" ? base.getIconName() : void 0) {
        return this.itemTitle.classList.add('icon', "icon-" + this.iconName);
      } else if (this.path != null && (this.iconName = FileIcons.getService().iconClassForPath(this.path, "tabs"))) {
        if (!Array.isArray(names = this.iconName)) {
          names = names.toString().split(/\s+/g);
        }

        return (ref2 = this.itemTitle.classList).add.apply(ref2, ['icon'].concat(slice.call(names)));
      }
    };

    TabView.prototype.isItemPending = function () {
      if (this.pane.getPendingItem != null) {
        return this.pane.getPendingItem() === this.item;
      } else if (this.item.isPending != null) {
        return this.item.isPending();
      }
    };

    TabView.prototype.terminatePendingState = function () {
      if (this.pane.clearPendingItem != null) {
        if (this.pane.getPendingItem() === this.item) {
          return this.pane.clearPendingItem();
        }
      } else if (this.item.terminatePendingState != null) {
        return this.item.terminatePendingState();
      }
    };

    TabView.prototype.clearPending = function () {
      this.itemTitle.classList.remove('temp');
      return this.element.classList.remove('pending-tab');
    };

    TabView.prototype.updateIconVisibility = function () {
      if (atom.config.get('tabs.showIcons')) {
        return this.itemTitle.classList.remove('hide-icon');
      } else {
        return this.itemTitle.classList.add('hide-icon');
      }
    };

    TabView.prototype.updateModifiedStatus = function () {
      var base;

      if (typeof (base = this.item).isModified === "function" ? base.isModified() : void 0) {
        if (!this.isModified) {
          this.element.classList.add('modified');
        }

        return this.isModified = true;
      } else {
        if (this.isModified) {
          this.element.classList.remove('modified');
        }

        return this.isModified = false;
      }
    };

    TabView.prototype.setupVcsStatus = function () {
      if (this.path == null) {
        return;
      }

      return this.repoForPath(this.path).then(function (_this) {
        return function (repo) {
          _this.subscribeToRepo(repo);

          return _this.updateVcsStatus(repo);
        };
      }(this));
    };

    TabView.prototype.subscribeToRepo = function (repo) {
      var ref1;

      if (repo == null) {
        return;
      }

      if ((ref1 = this.repoSubscriptions) != null) {
        ref1.dispose();
      }

      this.repoSubscriptions = new CompositeDisposable();
      this.repoSubscriptions.add(repo.onDidChangeStatus(function (_this) {
        return function (event) {
          if (event.path === _this.path) {
            return _this.updateVcsStatus(repo, event.pathStatus);
          }
        };
      }(this)));
      return this.repoSubscriptions.add(repo.onDidChangeStatuses(function (_this) {
        return function () {
          return _this.updateVcsStatus(repo);
        };
      }(this)));
    };

    TabView.prototype.repoForPath = function () {
      var dir, i, len, ref1;
      ref1 = atom.project.getDirectories();

      for (i = 0, len = ref1.length; i < len; i++) {
        dir = ref1[i];

        if (dir.contains(this.path)) {
          return atom.project.repositoryForDirectory(dir);
        }
      }

      return Promise.resolve(null);
    };

    TabView.prototype.updateVcsStatus = function (repo, status) {
      var newStatus;

      if (repo == null) {
        return;
      }

      newStatus = null;

      if (repo.isPathIgnored(this.path)) {
        newStatus = 'ignored';
      } else {
        if (status == null) {
          status = repo.getCachedPathStatus(this.path);
        }

        if (repo.isStatusModified(status)) {
          newStatus = 'modified';
        } else if (repo.isStatusNew(status)) {
          newStatus = 'added';
        }
      }

      if (newStatus !== this.status) {
        this.status = newStatus;
        return this.updateVcsColoring();
      }
    };

    TabView.prototype.updateVcsColoring = function () {
      this.itemTitle.classList.remove('status-ignored', 'status-modified', 'status-added');

      if (this.status && atom.config.get('tabs.enableVcsColoring')) {
        return this.itemTitle.classList.add("status-" + this.status);
      }
    };

    TabView.prototype.unsetVcsStatus = function () {
      var ref1;

      if ((ref1 = this.repoSubscriptions) != null) {
        ref1.dispose();
      }

      delete this.status;
      return this.updateVcsColoring();
    };

    return TabView;
  }();
}).call(void 0);
